src/Models/DomesticAddress.php

```diff
@@ -4,29 +4,72 @@
 
 namespace Tipoff\Addresses\Models;
 
+use Assert\Assert;
+use Carbon\Carbon;
 use Tipoff\Support\Models\BaseModel;
 use Tipoff\Support\Traits\HasPackageFactory;
 
+/**
+ * @property int id
+ * @property string address_line_1
+ * @property string address_line_2
+ * @property City city
+ * @property Zip zip
+ * @property Carbon created_at
+ * @property Carbon updated_at
+ * // Raw relations
+ * @property int city_id
+ * @property string zip_code
+ */
 class DomesticAddress extends BaseModel
 {
     use HasPackageFactory;
 
-    protected $casts = [];
+    protected $casts = [
+        'id' => 'integer',
+        'city_id' => 'integer',
+    ];
+
+    protected $fillable = [
+        'address_line_1',
+        'address_line_2',
+        'city_id',
+        'zip_code',
+    ];
+
+    /**
+     * @param string $line1
+     * @param string|null $line2
+     * @param string|City $city
+     * @param string|Zip $zip
+     * @return static
+     */
+    public static function createDomesticAddress(string $line1, ?string $line2, $city, $zip): self
+    {
+        $zip = ($zip instanceof Zip) ? $zip : Zip::query()->findOrFail(trim($zip));
+        $city = ($city instanceof City) ? $city : City::query()->byTitle($city)->firstOrFail();
+
+        /** @var DomesticAddress $domesticAddress */
+        $domesticAddress = static::query()->firstOrCreate([
+            'address_line_1' => trim($line1),
+            'address_line_2' => empty($line2) ? null : trim($line2),
+            'city_id' => $city->id,
+            'zip_code' => $zip->code,
+        ]);
+
+        return $domesticAddress;
+    }
 
     protected static function boot()
     {
         parent::boot();
 
-        static::saving(function ($address) {
-            if (empty($address->address_line_1)) {
-                throw new \Exception('US domestic addresses must have a street.');
-            }
-            if (empty($address->city_id)) {
-                throw new \Exception('US domestic addresses must have a city.');
-            }
-            if (empty($address->zip_code)) {
-                throw new \Exception('US domestic addresses must have a zip code.');
-            }
+        static::saving(function (DomesticAddress $address) {
+            Assert::lazy()
+                ->that($address->address_line_1)->notEmpty('US domestic addresses must have a street.')
+                ->that($address->city_id)->notEmpty('US domestic addresses must have a city.')
+                ->that($address->zip_code)->notEmpty('US domestic addresses must have zip code.')
+                ->verifyNow();
         });
```

Comment:
This is so much cleaner! Thank you!

Reply:
@chx2 Could you go through all the other models in the packages where I use this style of exception messages and refactor to how it is done here?